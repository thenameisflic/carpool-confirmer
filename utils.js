import { StackActions, NavigationActions } from "react-navigation";

export const createResetAction = navigationOpts => {
  return StackActions.reset({
    index: 0,
    actions: [NavigationActions.navigate(navigationOpts)]
  });
};

export const mockFetch = response => {
  const mockSuccessResponse = response;
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  return Promise.resolve({
    json: () => mockJsonPromise
  });
};
