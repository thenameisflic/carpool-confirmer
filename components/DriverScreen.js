import React from "react";
import { View, StyleSheet, Image, Button } from "react-native";
import Title from "./typography/Title";
import theme from "../theme";
import RoleIcon from "./RoleIcon";
import RoleToggler from "./RoleToggler";
import { createResetAction } from "../utils";
import { ROLES } from "../constants";

export default function DriverScreen({ navigation }) {
  return (
    <View style={styles.container}>
      <Title size="h5" style={styles.title}>
        Peça para o seu passageiro escanear o código abaixo
      </Title>
      <Image
        source={{ uri: "https://placehold.it/300x300" }}
        style={styles.qrCode}
      />
      <Title size="h5" style={styles.driverId}>
        Seu id de motorista é
      </Title>
      <Title size="h6" style={styles.textAccent}>
        hardcore-blackwell
      </Title>
      <Button
        title="TODO: Skip"
        onPress={() =>
          navigation.dispatch(
            createResetAction({
              routeName: "ShareLocation",
              params: { role: ROLES.DRIVER }
            })
          )
        }
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: theme.backgroundColor,
    paddingHorizontal: 16,
    alignItems: "center",
    justifyContent: "center",
    flex: 1
  },
  title: {
    textAlign: "center"
  },
  qrCode: {
    width: 300,
    height: 300,
    marginTop: 24
  },
  driverId: {
    textAlign: "center",
    marginTop: 24
  },
  textAccent: {
    color: theme.primaryColor
  },
  headerLeft: {
    marginLeft: 8
  },
  headerRight: {
    paddingRight: 8
  }
});

DriverScreen.navigationOptions = ({ navigation }) => ({
  title: "Motorista",
  headerLeft: <RoleIcon style={styles.headerLeft} />,
  headerStyle: {
    elevation: 0,
    shadowOpacity: 0
  },
  headerRight: (
    <View style={styles.headerRight}>
      <RoleToggler
        style={styles.headerRight}
        onPress={() => navigation.navigate("Passenger")}
      />
    </View>
  )
});
