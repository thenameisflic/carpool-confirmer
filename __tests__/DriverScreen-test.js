import "react-native";
import React from "react";
import DriverScreen from "../components/screens/DriverScreen";

import renderer from "react-test-renderer";
import { mockFetch, createResetAction } from "../utils";

it("renders DriverScreen correctly", () => {
  const tree = renderer.create(<DriverScreen navigation={{navigate: ""}} />);
  expect(tree.toJSON()).toMatchSnapshot();
});

it("doesn't let the user proceed if the passenger hasn't already scanned the code", () => {
    fetch = jest.fn(() => mockFetch({
        passengerScanned: false
    }));
    const navigate = jest.fn();
    const instance = renderer.create(<DriverScreen navigation={{navigate}} />);
    const button = instance.root.findByProps({title: "Continuar"});
    button.props.onPress();
    expect(fetch).toBeCalled();
    expect(navigate).not.toHaveBeenCalled();
});
  
test("navigates away if the user has scanned the code", async () => {
    fetch = jest.fn(() => mockFetch({
        passengerScanned: true,
        id: 1,
        driverPosition: "-10,-10",
        passengerPosition: "-10,-10",
        distance: 10
    }));
    const dispatch = jest.fn();
    const instance = renderer.create(<DriverScreen navigation={{dispatch}} />);
    const button = instance.root.findByProps({title: "Continuar"});
    button.props.onPress();
    await((await fetch()).json());
    expect(dispatch).toBeCalledWith(expect.objectContaining(createResetAction({
        routeName: "ShareLocation",
        params: expect.objectContaining({
            driverPosition: "-10,-10",
            passengerPosition: "-10,-10",
            distance: 10,
            role: "DRIVER"
        })
    })));
});
  