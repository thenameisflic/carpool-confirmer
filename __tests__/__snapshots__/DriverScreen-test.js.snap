// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DriverScreen correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 16,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Nunito Regular",
          "fontSize": 24,
        },
        Object {
          "textAlign": "center",
        },
      ]
    }
  >
    Peça para o seu passageiro escanear o código abaixo
  </Text>
  <View
    style={
      Object {
        "height": 300,
        "marginTop": 24,
        "width": 300,
      }
    }
  />
  <View
    accessible={true}
    clickable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "stretch",
        "backgroundColor": "#FF5E57",
        "borderRadius": 4,
        "height": 48,
        "justifyContent": "center",
        "marginTop": 24,
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "Nunito Bold",
            "fontSize": 20,
          },
          Object {
            "color": "#fff",
          },
        ]
      }
    >
      Continuar
    </Text>
  </View>
</View>
`;
